src/units.rs

```diff
@@ -0,0 +1,385 @@
+    }
+
+    pub(crate) fn coord(&self) -> UnitCoord {
+        (self.unit.inner.round(), self.unit.inner.creator()).into()
+    }
+
+    pub(crate) fn round(&self) -> usize {
+        self.unit.inner.round()
+    }
+
+    pub(crate) fn creator(&self) -> NodeIndex {
+        self.unit.inner.creator()
+    }
+
+    pub(crate) fn n_parents(&self) -> NodeCount {
+        self.unit.inner.n_parents()
+    }
+
+    pub(crate) fn n_members(&self) -> NodeCount {
+        self.unit.inner.n_members()
+    }
+
+    pub(crate) fn sign<KB: KeyBox<Signature>>(
+        keybox: &KB,
+        unit: FullUnit<H, D>,
+    ) -> SignedUnit<H, D, Signature> {
+        let encoded = unit.encode();
+        let signature = keybox.sign(&encoded);
+
+        SignedUnit {
+            unit,
+            signature,
+            id: keybox
+                .my_index()
+                .expect("Consensus is run only by validators"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Encode, Decode, StdHash)]
+pub(crate) struct UnitCoord {
+    pub(crate) creator: NodeIndex,
+    pub(crate) round: u64,
+}
+
+impl<H: Hash> From<Unit<H>> for UnitCoord {
+    fn from(unit: Unit<H>) -> Self {
+        UnitCoord {
+            creator: unit.creator(),
+            round: unit.round() as u64,
+        }
+    }
+}
+
+impl<H: Hash> From<&Unit<H>> for UnitCoord {
+    fn from(unit: &Unit<H>) -> Self {
+        UnitCoord {
+            creator: unit.creator(),
+            round: unit.round() as u64,
+        }
+    }
+}
+
+impl From<(usize, NodeIndex)> for UnitCoord {
+    fn from(coord: (usize, NodeIndex)) -> Self {
+        UnitCoord {
+            creator: coord.1,
+            round: coord.0 as u64,
+        }
+    }
+}
+
+type UnitRound = u64;
+
+#[derive(Clone, Debug, Default, PartialEq, Encode, Decode)]
+pub(crate) struct PreUnit<H: Hash> {
+    pub(crate) creator: NodeIndex,
+    pub(crate) round: UnitRound,
+    pub(crate) control_hash: ControlHash<H>,
+}
+
+impl<H: Hash> PreUnit<H> {
+    pub(crate) fn creator(&self) -> NodeIndex {
+        self.creator
+    }
+
+    pub(crate) fn round(&self) -> Round {
+        self.round as Round
+    }
+
+    pub(crate) fn new_from_parents(
+        creator: NodeIndex,
+        round: Round,
+        parents: NodeMap<Option<H>>,
+        hashing: impl Fn(&[u8]) -> H,
+    ) -> Self {
+        let control_hash = ControlHash::new(&parents, &hashing);
+        PreUnit {
+            creator,
+            round: round as u64,
+            control_hash,
+        }
+    }
+
+    pub(crate) fn n_parents(&self) -> NodeCount {
+        self.control_hash.n_parents()
+    }
+
+    pub(crate) fn n_members(&self) -> NodeCount {
+        self.control_hash.n_members()
+    }
+}
+
+impl<H: Hash> From<PreUnit<H>> for NotificationOut<H> {
+    fn from(pu: PreUnit<H>) -> Self {
+        NotificationOut::CreatedPreUnit(pu)
+    }
+}
+
+#[derive(Clone, Debug, Default, PartialEq, Encode, Decode)]
+pub(crate) struct Unit<H: Hash> {
```

Comment:
I think fields through template parameters would mostly make things less readable, especially while reading `unit.extra` in two places and having to figure out why they do such different things.

The `Signed` idea is imo great though – it would even be useful for some other data types (alerts? maybe anything one sends over RMC?), and we could probably implement it generally enough to abstract away all the signing and verification logic there.

Reply:
I think fields through template parameters would mostly make things less readable, especially while reading `unit.extra` in two places and having to figure out why they do such different things.

The `Signed` idea is imo great though – it would even be useful for some other data types (alerts? maybe anything one sends over RMC?), and we could probably implement it generally enough to abstract away all the signing and verification logic there.

Reply:
I think this is worth considering, but I its another thing that can be done as a separate task/separate PR.